### DISCUSSION TITLE
Postgres: defaultRaw not null throws error

### DISCUSSION THREAD

**Discussion by micchickenburger** [mikro-orm/mikro-orm#3883] [Q&A]
**Describe the bug**

Consider the following model:

```typescript
@Entity()
export default class Book {
  @Property({ defaultRaw: 'now()', nullable: false })
    timestamp?: Date;
}
```

The `timestamp` property is optional because we want to be able to instantiate `Book` without ever supplying `timestamp`, which is populated by the database.  The table creation SQL looks like this:

```sql
create table "book" ("timestamp" timestamptz(0) not null default now());
```

However, creating a new Book will execute the following SQL:

```sql
insert into "book" ("timestamp") values (NULL);
```

This throws an error due to the `NOT NULL` constraint.  It seems that for Postgres an undefined property with a default database value should send `DEFAULT` instead of `NULL`, or not specify the column at all.

**Expected behavior**
The database should be able to populate this field without specifying a value during instantiation.

**Additional context**
See more here:  https://stackoverflow.com/questions/67013313/want-not-null-and-default-to-false-if-supplied-value-is-null

**Versions**

| Dependency | Version |
| - | - |
| node | 16.19.0 |
| typescript | 4.8.3 |
| mikro-orm | 5.6.3 |
| postgres | 5.6.3 |

**Comment by B4nan** [comment]
I dont see any code reproduction, so this feels like you are complaining about how SQL works? Or what code did you use that produced the failing query? You apparently had to pass the null value yourself somewhere, otherwise it won't be there.

**Reply by micchickenburger** [reply]
```typescript
const book = new Book();
em.persist(book);
await em.flush();
```

I'm not specifying a value anywhere.

**Reply by B4nan** [reply]
I will need to see a complete reproduction, there are many tests that does this.

Here is a passing test case:

```ts
import { Entity, MikroORM, PrimaryKey, Property } from '@mikro-orm/core';
import { PostgreSqlDriver } from '@mikro-orm/postgresql';

@Entity()
export class Test {

  @PrimaryKey()
  id!: number;

  @Property({ defaultRaw: 'now()', nullable: false })
  timestamp?: Date;

}

describe('GH issue 3883', () => {

  let orm: MikroORM;

  beforeAll(async () => {
    orm = await MikroORM.init({
      entities: [Test],
      dbName: 'mikro_orm_test_2489',
      driver: PostgreSqlDriver,
      debug: true,
    });
    await orm.schema.refreshDatabase();
  });

  afterAll(() => orm.close(true));

  test('3883', async () => {
    const e = new Test();
    await orm.em.persist(e).flush();
  });

});
```

**Reply by micchickenburger** [reply]
Ah, my apologies, I thought it would be a more general case.  It happens with embeddables.

```typescript
@Embeddable()
abstract class Time {
  @Property({ defaultRaw: 'now()', nullable: false })
    timestamp?: Date;
}

@Entity()
class Test extends BaseEntity<Test> {
  @Embedded()
    time!: Time;
}

await dbConnection.open();
const em = dbConnection.em?.fork();
if (em) {
  const e = new Test();
  e.time = {};
  await em.persist(e).flush();
}
```

```
NotNullConstraintViolationException: insert into "test" ("id", "time_timestamp") values ('50a071d3-633e-4055-9e0d-edbe44a0e085', NULL) returning "id" - null value in column "time_timestamp" of relation "test" violates not-null constraint
```

**Reply by B4nan** [reply]
I see, embeddables have quite a lot of restrictions (that are being gradually lifted), because they can be used in object mode (= JSON properties) as well as nested inside each other, or even as arrays (again, JSON properties). Those places will never support things like database-level defaults or FK constraints.

When I designed the feature initially, I totally missed one use case that is popping up more often these days. The inline mode embeddables can be seen as some kind of a PHP trait, or a partial if you want. In that sense (and in the context of inline mode usage, even with nesting) you could use any kind of property in there, including primary keys or collection properties. We would have to validate carefully that such properties/features are only used on inline mode, but other than that, it might work just fine.

Supporting this might take a bit more effort than you might think, but it's something I would like to revisit in v6 (development is already in progress). Right now I would suggest using runtime defaults, I'd say it's generally a better solution anyway - you get the value even before the entity reaches the database. That way you don't have to mark your `timestamp` property as optional, which also sounds more accurate to me, as you always have a value.

```ts
  @Property({ defaultRaw: 'now()', nullable: false })
    timestamp: Date = new Date();
```

**Reply by micchickenburger** [reply]
Thank you, that makes sense.  My timestamp field is actually called `recorded`, so I wanted it to be the time in which the entity was added to the database.  I store timestamps for various events for performance analysis down the road.
